Constraint generation is higher order now.

awruef opened the pull request:
Changed to a squash merge.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -12,6 +12,125 @@
+// If they are both FVConstraint, then do a return-value and parameter
+// by parameter constraint generation.
+// If they are of an unequal parameter type, constrain everything in both
+// to top.
```
>top. [](start = 6, length = 4)

Use the term wild instead?

dtarditi commented on tools/checked-c-convert/ProgramInfo.cpp:
```diff
@@ -8,52 +8,313 @@
+        } else {
+          s = s + BaseType + "*";
+        }
+      }
```
We agree that this probably won't work where any of the constraints on the function pointer are not PTR.   You should probably make the code not do rewrites (for now) in the cases that will go wrong.

awruef replied:
I added a test to `PointerVariableConstraint::anyChanges` so that the re-write won't happen in that case.

dtarditi commented on tools/checked-c-convert/ProgramInfo.h:
```diff
@@ -33,80 +33,140 @@
+  PointerVariableConstraint(CVars V, std::string T) : 
+    ConstraintVariable(PointerVariable, T),vars(V),FV(nullptr) {}
+  PointerVariableConstraint(clang::DeclaratorDecl *D, uint32_t &K,
+    Constraints &CS);
```
Please add a comment explaining what this function does.

dtarditi commented on tools/checked-c-convert/ProgramInfo.h:
```diff
@@ -120,15 +180,16 @@ class ProgramInfo {
 
   bool getDeclStmtForDecl(clang::Decl *D, clang::DeclStmt *&St);
 
   // Checks the structural type equality of two constraint variables. This is 
```
>two constraint variables. [](start = 44, length = 25)

The arguments are now sets, now individual variables.   What does this do for the the sets of variables?

awruef replied:
Right now, nothing, since `checkStructuralEquality` always returns `false`. When it starts returning something other than `false`, it will probably have to perform the comparison on the LUB of `V` and `U`.

dtarditi commented on test/CheckedCRewriter/function_tests.c:
```diff
@@ -0,0 +1,52 @@
+// Tests for Checked C rewriter tool.
```
How about tests for parameters that take function pointer types, returning a function pointer type, and arrays of function pointer types?

awruef replied:
I don't think arrays of function pointer types will work yet but I'll add it in the next branch.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -12,6 +12,125 @@
+      CLHS->dump();
+      errs() << "\n";
+      // Constrain everything in both to top.
+      llvm_unreachable("TODO");
```
Could you file an issue about the "TODO" work in this section?

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -57,51 +178,75 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
   // for each pair of q_i to q_j \forall j in variables_on_rhs.
-  void constrainAssign(uint32_t V, Expr *RHS) {
+  void constrainAssign( std::set<ConstraintVariable*> V, 
+                        Expr *RHS) {
```
Please add an explanation of what V is.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -57,51 +178,75 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
-                                        Expr::NPC_ValueDependentIsNotNull))
-          CS.addConstraint(CS.createEq(CS.getOrCreateVar(V), CS.getWild()));
+          Expr::NPC_ValueDependentIsNotNull))
+          for (const auto &U : W)
```
I don't understand why the iteration is over W.   I would think that it should be over V.

awruef replied:
That makes sense, changing.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -152,96 +291,88 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
+    // Constrain the value returned (if present) against the return value
+    // of the function.   
+    for (const auto &F : Fun )
+      if (FVConstraint *FU = dyn_cast<FVConstraint>(F))
```
>FU [](start = 24, length = 2)

Probably not a good abbbreviation for a variable name.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -152,96 +291,88 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
 
-  bool VisitUnaryPreInc(UnaryOperator *O) {
-    std::set<uint32_t> V;
+  void constrainExprFirst(Expr *E) {
```
>void constrainExprFirst(Expr *E) { [](start = 2, length = 34)

Please add a comment explaining what this function does.

dtarditi commented:
This is a nice generalization of the existing code.  Thanks for addressing my first set of comments.  I've finished the rest of my first rounds of reviews.   I"ve made some requests for additional comments and have a few questions about the code.

awruef commented:
I've addressed most of the feedback. For now we don't do any re-writing in "extreme" function pointer cases like arrays of function pointers. I'll write tests for those and other cases next.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -178,6 +177,9 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
   // variables on the RHS could be 0 or more. We just do the same rule
   // for each pair of q_i to q_j \forall j in variables_on_rhs.
+  //
+  // V is the set of constraint variables on the left hand side that we are
```
Could you amplify this comment some more?   It's the set of constraint variables for the LHS.    The LHS will have a pointer type and the constraint variables represent the constraints for that type.

dtarditi commented on tools/checked-c-convert/ConstraintBuilder.cpp:
```diff
@@ -57,51 +177,78 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
+  //
+  // V is the set of constraint variables on the left hand side that we are
+  // assigning to.
+  void constrainAssign( std::set<ConstraintVariable*> V, 
```
Could you amplify this comment some more? It's the set of constraint variables for the LHS. The LHS will have a pointer type and the constraint variables represent the constraints for that type.
